File: sdk/src/envelope.ts
/**
 * Proof envelope wire format: application/vnd.bolyra.proof+json
 *
 * Self-describing envelope for ZKP proofs with typed fields,
 * circuit identity binding, and version negotiation.
 */

/** Content-Type for HTTP headers. */
export const CONTENT_TYPE = 'application/vnd.bolyra.proof+json';

/** Current envelope version. */
export const ENVELOPE_VERSION = '1.0.0';

/** Valid circuit names. */
export type CircuitName = 'HumanUniqueness' | 'AgentPolicy' | 'Delegation';

/** Valid proof systems. v1 supports groth16 only (PLONK has different shape). */
export type ProofType = 'groth16';

/** Circuit identity with version and vkey binding. */
export interface CircuitIdentity {
  name: CircuitName;
  version: string;
  vkeyHash?: string; // sha256:<64 lowercase hex chars>
}

/** Groth16 proof coordinates. All values are decimal strings. */
export interface ProofData {
  pi_a: [string, string];
  pi_b: [[string, string], [string, string]];
  pi_c: [string, string];
}

/** Informational metadata. Verifiers MUST NOT reject based on these. */
export interface ProofMetadata {
  prover?: string;
  timestamp?: string; // RFC 3339
  [key: string]: unknown; // forward-compatible
}

/** The canonical proof envelope. */
export interface ProofEnvelope {
  version: string;
  circuit: CircuitIdentity;
  proofType: ProofType;
  publicSignals: string[];
  proof: ProofData;
  metadata?: ProofMetadata;
  [key: string]: unknown; // forward-compatible: unknown top-level fields preserved
}

// BN254 scalar field modulus (defined locally to avoid circular deps with identity.ts)
const BN254_FIELD_ORDER = 21888242871839275222246405745257275088548364400416034343698204186575808495617n;

const VALID_CIRCUITS: ReadonlySet<string> = new Set(['HumanUniqueness', 'AgentPolicy', 'Delegation']);
const VALID_PROOF_TYPES: ReadonlySet<string> = new Set(['groth16']); // v1: groth16 only

/**
 * Validate a decimal field element string.
 * - Rejects strings > 78 chars (DoS prevention before BigInt parsing)
 * - Rejects leading zeros (except "0" itself)
 * - Rejects values >= BN254 field modulus
 */
function validateFieldElement(s: string, label: string): void {
  if (typeof s !== 'string') {
    throw new Error(`${label}: must be a string, got ${typeof s}`);
  }
  // Reject strings > 78 chars before BigInt parsing (DoS prevention)
  if (s.length > 78) {
    throw new Error(`${label}: string too long (${s.length} chars, max 78)`);
  }
  // No leading zeros except "0" itself
  if (!/^(0|[1-9]\d*)$/.test(s)) {
    throw new Error(`${label}: must be a decimal string without leading zeros, got ${JSON.stringify(s)}`);
  }
  const n = BigInt(s);
  if (n >= BN254_FIELD_ORDER) {
    throw new Error(`${label}: value >= BN254 field modulus`);
  }
}

/** Parse a semver string into [major, minor, patch]. */
function parseSemver(v: string): [number, number, number] {
  const m = v.match(/^(\d+)\.(\d+)\.(\d+)$/);
  if (!m) throw new Error(`Invalid semver: ${v}`);
  return [parseInt(m[1], 10), parseInt(m[2], 10), parseInt(m[3], 10)];
}

/**
 * Check envelope version against the current supported version.
 * Major mismatch = reject. Minor/patch mismatch = accept (forward-compat).
 */
function checkVersion(envelopeVersion: string): void {
  const [major] = parseSemver(envelopeVersion);
  const [currentMajor] = parseSemver(ENVELOPE_VERSION);
  if (major !== currentMajor) {
    throw new Error(
      `Incompatible envelope version: ${envelopeVersion} (current: ${ENVELOPE_VERSION}). Major version mismatch.`
    );
  }
}

/**
 * Validate a raw object as a ProofEnvelope.
 * Throws on any structural or value violation.
 * Unknown top-level fields are preserved (forward-compat).
 */
export function validateEnvelope(envelope: Record<string, unknown>): ProofEnvelope {
  // Version
  if (typeof envelope.version !== 'string') {
    throw new Error('Missing or invalid version');
  }
  checkVersion(envelope.version);

  // Circuit identity
  const circuit = envelope.circuit as Record<string, unknown> | undefined;
  if (!circuit || typeof circuit !== 'object' || Array.isArray(circuit)) {
    throw new Error('Missing or invalid circuit');
  }
  if (typeof circuit.name !== 'string' || !VALID_CIRCUITS.has(circuit.name)) {
    throw new Error(`Invalid circuit.name: ${circuit.name}`);
  }
  if (typeof circuit.version !== 'string') {
    throw new Error('Missing circuit.version');
  }
  if (circuit.vkeyHash !== undefined) {
    if (
      typeof circuit.vkeyHash !== 'string' ||
      !/^sha256:[0-9a-f]{64}$/.test(circuit.vkeyHash)
    ) {
      throw new Error(
        `Invalid circuit.vkeyHash: must be sha256:<64 lowercase hex chars>, got ${JSON.stringify(circuit.vkeyHash)}`
      );
    }
  }

  // Proof type
  if (typeof envelope.proofType !== 'string' || !VALID_PROOF_TYPES.has(envelope.proofType)) {
    throw new Error(`Invalid proofType: ${envelope.proofType}`);
  }

  // Public signals
  const signals = envelope.publicSignals;
  if (!Array.isArray(signals) || signals.length === 0) {
    throw new Error('publicSignals must be a non-empty array');
  }
  (signals as unknown[]).forEach((s, i) => validateFieldElement(s as string, `publicSignals[${i}]`));

  // Proof coordinates
  const proof = envelope.proof as Record<string, unknown> | undefined;
  if (!proof || typeof proof !== 'object' || Array.isArray(proof)) {
    throw new Error('Missing proof');
  }

  const pi_a = proof.pi_a as unknown[];
  if (!Array.isArray(pi_a) || pi_a.length !== 2) {
    throw new Error('proof.pi_a must be [string, string]');
  }
  pi_a.forEach((s, i) => validateFieldElement(s as string, `pi_a[${i}]`));

  const pi_b = proof.pi_b as unknown[];
  if (!Array.isArray(pi_b) || pi_b.length !== 2) {
    throw new Error('proof.pi_b must be [[s,s],[s,s]]');
  }
  for (let r = 0; r < 2; r++) {
    const row = pi_b[r] as unknown[];
    if (!Array.isArray(row) || row.length !== 2) {
      throw new Error(`proof.pi_b[${r}] must be [string, string]`);
    }
    row.forEach((s, c) => validateFieldElement(s as string, `pi_b[${r}][${c}]`));
  }

  const pi_c = proof.pi_c as unknown[];
  if (!Array.isArray(pi_c) || pi_c.length !== 2) {
    throw new Error('proof.pi_c must be [string, string]');
  }
  pi_c.forEach((s, i) => validateFieldElement(s as string, `pi_c[${i}]`));

  return envelope as unknown as ProofEnvelope;
}

/**
 * Serialize a ProofEnvelope to a JSON string.
 * Unknown top-level fields are included (forward-compat).
 */
export function serializeEnvelope(envelope: ProofEnvelope): string {
  return JSON.stringify(envelope);
}

/**
 * Deserialize and validate a JSON string into a ProofEnvelope.
 * Throws on parse errors or validation failures.
 */
export function deserializeEnvelope(json: string): ProofEnvelope {
  let parsed: unknown;
  try {
    parsed = JSON.parse(json);
  } catch (e) {
    throw new Error(`Failed to parse envelope JSON: ${(e as Error).message}`);
  }
  if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
    throw new Error('Envelope must be a JSON object');
  }
  return validateEnvelope(parsed as Record<string, unknown>);
}

/**
 * Wrap raw snarkjs proof output into a ProofEnvelope.
 * The resulting envelope passes validateEnvelope() if inputs are valid field elements.
 */
export function envelopeFromSnarkjsProof(
  circuitName: CircuitName,
  proof: { pi_a: string[]; pi_b: string[][]; pi_c: string[] },
  publicSignals: string[],
  options?: { circuitVersion?: string; vkeyHash?: string },
): ProofEnvelope {
  if (!Array.isArray(publicSignals) || publicSignals.length === 0) {
    throw new Error('publicSignals must be a non-empty array');
  }
  if (!VALID_CIRCUITS.has(circuitName)) {
    throw new Error(`Invalid circuit name: ${circuitName}`);
  }

  const envelope: ProofEnvelope = {
    version: ENVELOPE_VERSION,
    circuit: {
      name: circuitName,
      version: options?.circuitVersion ?? '0.4.0',
      ...(options?.vkeyHash ? { vkeyHash: options.vkeyHash } : {}),
    },
    proofType: 'groth16',
    publicSignals,
    proof: {
      pi_a: [proof.pi_a[0], proof.pi_a[1]],
      pi_b: [
        [proof.pi_b[0][0], proof.pi_b[0][1]],
        [proof.pi_b[1][0], proof.pi_b[1][1]],
      ],
      pi_c: [proof.pi_c[0], proof.pi_c[1]],
    },
    metadata: {
      prover: '@bolyra/sdk@0.5.1',
      timestamp: new Date().toISOString(),
    },
  };

  return envelope;
}

File: sdk-python/bolyra/envelope.py
"""Proof envelope wire format: application/vnd.bolyra.proof+json

Self-describing envelope for ZKP proofs with typed fields,
circuit identity binding, and version negotiation.
"""
from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

CONTENT_TYPE = "application/vnd.bolyra.proof+json"
ENVELOPE_VERSION = "1.0.0"

# BN254 scalar field modulus
BN254_FIELD_ORDER = 21888242871839275222246405745257275088548364400416034343698204186575808495617

VALID_CIRCUITS = frozenset({"HumanUniqueness", "AgentPolicy", "Delegation"})
VALID_PROOF_TYPES = frozenset({"groth16"})  # v1: groth16 only


def _validate_field_element(s: str, label: str) -> None:
    if not isinstance(s, str):
        raise ValueError(f"{label}: must be a string, got {type(s).__name__}")
    if len(s) > 78:
        raise ValueError(f"{label}: string too long ({len(s)} chars, max 78)")
    if not re.match(r"^(0|[1-9]\d*)$", s):
        raise ValueError(f"{label}: must be a decimal string without leading zeros, got {s!r}")
    n = int(s)
    if n >= BN254_FIELD_ORDER:
        raise ValueError(f"{label}: value >= BN254 field modulus")


def _parse_semver(v: str) -> tuple[int, int, int]:
    m = re.match(r"^(\d+)\.(\d+)\.(\d+)$", v)
    if not m:
        raise ValueError(f"Invalid semver: {v}")
    return int(m.group(1)), int(m.group(2)), int(m.group(3))


def _check_version(version: str) -> None:
    major, _, _ = _parse_semver(version)
    current_major, _, _ = _parse_semver(ENVELOPE_VERSION)
    if major != current_major:
        raise ValueError(
            f"Incompatible envelope version: {version} "
            f"(current: {ENVELOPE_VERSION}). Major version mismatch."
        )


@dataclass
class CircuitIdentity:
    name: str
    version: str
    vkey_hash: str | None = None

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {"name": self.name, "version": self.version}
        if self.vkey_hash:
            d["vkeyHash"] = self.vkey_hash
        return d

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> CircuitIdentity:
        return cls(
            name=d["name"],
            version=d["version"],
            vkey_hash=d.get("vkeyHash"),
        )


@dataclass
class ProofData:
    pi_a: list[str]
    pi_b: list[list[str]]
    pi_c: list[str]

    def to_dict(self) -> dict[str, Any]:
        return {"pi_a": self.pi_a, "pi_b": self.pi_b, "pi_c": self.pi_c}

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ProofData:
        return cls(pi_a=d["pi_a"], pi_b=d["pi_b"], pi_c=d["pi_c"])


@dataclass
class ProofEnvelope:
    version: str
    circuit: CircuitIdentity
    proof_type: str
    public_signals: list[str]
    proof: ProofData
    metadata: dict[str, Any] = field(default_factory=dict)
    extra: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "version": self.version,
            "circuit": self.circuit.to_dict(),
            "proofType": self.proof_type,
            "publicSignals": self.public_signals,
            "proof": self.proof.to_dict(),
        }
        if self.metadata:
            d["metadata"] = self.metadata
        d.update(self.extra)
        return d

    def to_json(self) -> str:
        return json.dumps(self.to_dict())

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ProofEnvelope:
        known_keys = {"version", "circuit", "proofType", "publicSignals", "proof", "metadata"}
        extra = {k: v for k, v in d.items() if k not in known_keys}
        return cls(
            version=d["version"],
            circuit=CircuitIdentity.from_dict(d["circuit"]),
            proof_type=d["proofType"],
            public_signals=d["publicSignals"],
            proof=ProofData.from_dict(d["proof"]),
            metadata=d.get("metadata", {}),
            extra=extra,
        )

    @classmethod
    def from_json(cls, raw: str) -> ProofEnvelope:
        d = json.loads(raw)
        return validate_envelope(d)


def validate_envelope(d: dict[str, Any]) -> ProofEnvelope:
    """Validate a raw dict and return a ProofEnvelope. Raises ValueError on invalid input."""
    if not isinstance(d.get("version"), str):
        raise ValueError("Missing or invalid version")
    _check_version(d["version"])

    circuit = d.get("circuit")
    if not isinstance(circuit, dict):
        raise ValueError("Missing or invalid circuit")
    if circuit.get("name") not in VALID_CIRCUITS:
        raise ValueError(f"Invalid circuit.name: {circuit.get('name')}")
    if not isinstance(circuit.get("version"), str):
        raise ValueError("Missing circuit.version")

    if d.get("proofType") not in VALID_PROOF_TYPES:
        raise ValueError(f"Invalid proofType: {d.get('proofType')}")

    signals = d.get("publicSignals")
    if not isinstance(signals, list) or len(signals) == 0:
        raise ValueError("publicSignals must be a non-empty array")
    for i, s in enumerate(signals):
        _validate_field_element(s, f"publicSignals[{i}]")

    proof = d.get("proof")
    if not isinstance(proof, dict):
        raise ValueError("Missing proof")
    for coord in ("pi_a", "pi_c"):
        arr = proof.get(coord)
        if not isinstance(arr, list) or len(arr) != 2:
            raise ValueError(f"proof.{coord} must be [string, string]")
        for i, s in enumerate(arr):
            _validate_field_element(s, f"{coord}[{i}]")
    pi_b = proof.get("pi_b")
    if not isinstance(pi_b, list) or len(pi_b) != 2:
        raise ValueError("proof.pi_b must be [[s,s],[s,s]]")
    for r, row in enumerate(pi_b):
        if not isinstance(row, list) or len(row) != 2:
            raise ValueError(f"proof.pi_b[{r}] must be [string, string]")
        for c, s in enumerate(row):
            _validate_field_element(s, f"pi_b[{r}][{c}]")

    return ProofEnvelope.from_dict(d)


def envelope_from_proof(
    circuit_name: str,
    proof: dict[str, Any],
    public_signals: list[str],
    *,
    circuit_version: str = "0.4.0",
    vkey_hash: str | None = None,
) -> ProofEnvelope:
    """Wrap raw proof output into a ProofEnvelope."""
    return ProofEnvelope(
        version=ENVELOPE_VERSION,
        circuit=CircuitIdentity(
            name=circuit_name,
            version=circuit_version,
            vkey_hash=vkey_hash,
        ),
        proof_type="groth16",
        public_signals=public_signals,
        proof=ProofData(
            pi_a=[proof["pi_a"][0], proof["pi_a"][1]],
            pi_b=[
                [proof["pi_b"][0][0], proof["pi_b"][0][1]],
                [proof["pi_b"][1][0], proof["pi_b"][1][1]],
            ],
            pi_c=[proof["pi_c"][0], proof["pi_c"][1]],
        ),
        metadata={
            "prover": "bolyra@0.5.0",
            "timestamp": datetime.now(timezone.utc).isoformat(),
        },
    )
